File: src/hackercup/squished.clj
(ns hackercup.squished
  (:require [clojure.java.io :as io]
            [clojure.string :as string]
            [clojure.math.combinatorics :as combinatorics]
            ))

(def input "/Users/tomh/Dropbox/Projects/hackercup/src/hackercup/squished-status-in.txt")
(def output "/Users/tomh/Dropbox/Projects/hackercup/src/hackercup/squished-status-out.txt")

(defn process-file [f input output]
  (with-open [output-file (io/writer output)
              input-file (io/reader input)]
    (doseq [output-line (map (partial format "Case #%d: %s\n" )
                             (iterate inc 1)
                             (map #(apply f %) (partition 2 (drop 1 (string/split (slurp input-file)  #"\s+" )))))]
        (do (.write output-file output-line)
            (print output-line)
            ))))

(defn bigparse [somestring]
  (new BigInteger somestring))

(defn to-int [str-partitions]
  (map (fn [l] (map #(bigparse (apply str %)) l)) str-partitions))

(defn remove-zeros [str-partitions]
  (remove (fn [l]
            (some #(= \0 %)
                  (map first l)))
          str-partitions))
  
(defn partitions [someseq]
  (map (partial partition-by-seq someseq)
       (lazy-cat ()))))

(defn bigparse [somestring]
  (new BigInteger somestring))

(defn squished [M message]
  (let [valid-nums (set (range (inc (bigparse M))))
        str-partitions (partitions message)
        possible-messages (count
                           (filter
                            #(every? valid-nums %)
                            (to-int
                            (remove-zeros
                             str-partitions)))
        )
    ]
    (mod possible-messages 4207849484)))

(defn -main []
  (process-file squished input output))

File: src/hackercup/squished.py
import math
import random
def partitions(somestring,valid_chars,sofar):
	l = len(somestring)
	if l == 0:
		yield sofar
	one = somestring[0]
	if one in valid_chars:
		partitions(somestring[1:],valid_chars,[one]+sofar)
	if l >= 2:
		two = somestring[1]
		if one+two in valid_chars:
			partitions(somestring[2:],valid_chars,[one+two]+sofar)
	if l >= 3:
		three = somestring[2]
		if one+two+three in valid_chars:
			partitions(somestring[3:],valid_chars,[one+two+three] + sofar)


test_str = ''
for i in range(100):
	test_str += str(int(math.ceil(random.random() * 10000000000)))

print(len(test_str))
count = 0
for i in partitions('1111111111111222222',['1','2','12'],[]):
	print i
	count += 1
	

print(count)
	
	
